airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java

```diff
@@ -127,16 +133,18 @@ public void prepareTables() throws Exception {
   public void typeAndDedupe(final String originalNamespace, final String originalName) throws Exception {
     LOGGER.info("Attempting typing and deduping for {}.{}", originalNamespace, originalName);
     final var streamConfig = parsedCatalog.getStream(originalNamespace, originalName);
-    if (streamsWithSuccesfulSetup.stream()
-        .noneMatch(streamId -> streamId.originalNamespace().equals(originalNamespace) && streamId.originalName().equals(originalName))) {
-      // For example, if T+D setup fails, but the consumer tries to run T+D on all streams during close,
-      // we should skip it.
-      LOGGER.warn("Skipping typing and deduping for {}.{} because we could not set up the tables for this stream.", originalNamespace, originalName);
-      return;
+    synchronized(tdLocks.get(streamConfig.id())) {
```

Comment:
good point. synchronized means they'll all wait their turn. probably need to use a stronger concurrency object here.

we'll still need a way to force T+D to run at the end of a sync (i.e. using the wait-my-turn behavior). There's probably something fancy to do there, e.g. track the last time we started a T+D run vs the last time we committed raw data and skip if there's no new data... but that can be a future improvement

Reply:
> we'll still need a way to force T+D to run at the end of a sync
for some reason I thought we hooked into the shutdown behavior for that 🤷

Reply:
> shutdown behavior

we do. I think I was just mistaken about this - was worried about a case where we have an in-flight T+D run while the shutdown hook is running, which I don't think is possible.

Reply:
added an explicit `mustRun` param anyway, because I was having a hard time writing a comment explaining why it isn't necessary :shrug: and if anyone ever gets clever and runs T+D in a separate thread, then we probably need it anyway